### DISCUSSION TITLE
How to achieve linear interpolation required by casadi structure?

### DISCUSSION THREAD

**Discussion by zmdntnu** [do-mpc/do-mpc#219] [Q&A] [answered]
Hi, @4flixt, here is my another big issue. 

In my case, alfa_eff = delta - gamma, where delta is control variable (_u(1x1)), gamma is calculated using state variables (_x(3x1), and I would like to use it to calculate the interpolation value of a pre-defined interpolation function. But I don't know which function to use. I checked the casadi guidance, and found casadi.interpolant. Here are the example codes I wrote:

    alfa_eff = delta - gamma
    inflow_ang_grid = np.array([0, 0.09, 0.17, 0.26, 0.35, 0.44, 0.52, 0.61, 0.7, 0.79, 1.2])
    C_l_grid = [1,2,3,3,4,3,3,2,3,2,1]
    C_d_grid =  [1,2,3,3,4,3,3,2,3,2,1]
    C_l_p =  casadi.interpolant('C_l_p','linear',[inflow_ang_grid],C_l_grid)
    C_l = sign(alfa_eff) * C_l_p(fabs(alfa_eff))
    C_d_p = casadi.interpolant('C_d_p','linear',[inflow_ang_grid],C_d_grid)
    C_d = C_d_p(fabs(alfa_eff))

But this leads to 
Traceback (most recent call last):
  File "D:/python file'/MPC_planning/example_4.py", line 176, in <module>
    X_w = C_X_p(fabs(epsilon)) * 1.23 / 2 * Ua**2 * A_F
  File "D:\python file'\venv\lib\site-packages\casadi\casadi.py", line 13444, in __call__
    ret = self.call(args)
  File "D:\python file'\venv\lib\site-packages\casadi\casadi.py", line 12324, in call
    return _casadi.Function_call(self, *args)
RuntimeError: .../casadi/core/function_internal.cpp:1832: 'eval_sx' not defined for LinearInterpolant

I guess it's probably because that I didn't use the correct interpolation function. 

BR

**Comment by 4flixt** [comment, marked as answer]
Hi,

you cannot use lookup tables in symbolic operations. See the [example](https://web.casadi.org/docs/#d-lookup-tables) from the CasADi website:

```python
import casadi as ca

xgrid = np.linspace(1,6,6)
V = [-1,-1,-2,-3,0,2]
lut = ca.interpolant('LUT','bspline',[xgrid],V)
print(lut(2.5))
```

Evaluation with NUMERICAL value 2.5 works fine and yields:
```
>> 1.35
```

Now the same with symbolic variable:
```python
x = ca.SX.sym('x')

lut = ca.interpolant('LUT','bspline',[xgrid],V)
print(lut(x))
```
Returns:
```
RuntimeError: .../casadi/core/function_internal.cpp:1832: 'eval_sx' not defined for BSplineInterpolant
```

Without symbolic evaluation of a computational graph, CasADi and thus do-mpc cannot compute fast gradient information (with automatic differentiation) and thus isnt able to solve the MPC optimization problem (fast).

Best,
4flixt

**Reply by zmdntnu** [reply]
Dear @4flixt, 

Thank you so much for your reply. So this is the dead end and there's no way out. (I don't know why they have to make the ship model like this. It's so difficult to use.)

Best wishes

zmdntnu

**Reply by alanliowh** [reply]
Hi @4flixt and @zmdntnu,

There is a workaround to use a look-up table. What you need to do is to initialise the model with `MX` symbolics. E.g.

`model = do_mpc.model.Model('continuous','MX')`.

Somehow CasADi interpolant can only handle symbolic variables defined in MX symbolics, not SX symbolics.

Best regards,
alanliowh

**Reply by 4flixt** [reply]
Hi @alanliowh ,

thank you for the suggestion!

Best,
4flixt

**Reply by pcesare** [reply]
@alanliowh I apologize but I'm a beginner, could you explain how to use the interpolant with casadi symbolics, if possible? I have a constraint that calls an interpolant function and I get the error: 'eval_sx' not defined for LinearInterpolant. The constraint is:

for i=1:N-1
opti.subject_to(OCV_sym(:,i+1) == casadinterp(Speed_sym(:,i), OCV_sym(:,i), Irr(i)));
end

with speed and OCV being MX symbolics. Thanks a lot
